Search: lift out searcher callout in frontend.

rvantonder opened the pull request:
- It's useful to have a generic function that does "run all of these repos through the searcher endpoint", for an arbitrary list of repos. 
- Right now, such logic is inlined in `searchFilesinRepos` and only does "run all of these (unindexed repos) through searcher endpoint". 
- I want to preserve this logic, but also make it possible to "run all of these <repos/files zoekt tells me about> through the searcher endpoint", for structural search.

Such a function cannot be extracted to the top level because it runs goroutines that rely on `addMatches` (which seeks to be race safe), and is also inlined. This PR extracts the desired logic into a dedicated inline function `callSearcherOverRepos`, which I will use in subsequent diffs to call out to searcher for structural search.

 What I need is for Go experts to confirm that what I've done here is semantics preserving and race safe.

keegancsmith commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -530,6 +530,92 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
+				addMatches(matches)
+			}(limitCtx, limitDone, repoRev)
+		}
+		wg.Wait()
```
you probably want this `wg.Wait` to stay outside of this lifting, since that applies to other goroutines like zoekt.

rvantonder replied:
That makes a whole lot of sense!

rvantonder commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -530,6 +530,91 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
 		}
 	}
 
+	// This  function calls searcher on a set of repos.
```
Good eye!

rvantonder commented:
Pulling out `wg.Wait()` breaks tests :grimacing: There's a [data race dump](https://buildkite.com/sourcegraph/sourcegraph/builds/46277#8d51fb5b-2238-42e6-b6af-343954de2deb/165-166) in the build, will debug why (maybe the test needs to be updated).

> [2019-10-25T03:00:51Z]     textsearch_test.go:358: searching non-existent rev expected to fail with RevisionNotFoundError got: <nil> --
> | [2019-10-25T03:00:51Z]     testing.go:853: race detected during execution of test

unknwon commented:
As a quick look, two of two are caused by reusing the `err` variable. Because we have named return value `err error`, `err` inside goroutine accidentally points to it.

I am not sure was this intentional, also the code before this PR will have race condition for the same reason, but the tests somehow didn't catch it.

---

This `searchFilesInRepos` is crazy long!

rvantonder commented:
Ah ha. Thanks for taking a peek 🙏. I'll fix up the tests.

> This searchFilesInRepos is crazy long!

Oh definitely... that's why I'm trying to tread carefully :P It's not easy to simplify right now with all the activity going on there.

unknwon commented:
Hmm, these two races look totally valid to me, we should instead fix the code not tests?

rvantonder commented:
Oh oops, I misunderstood and didn't realize this was in reference to the code and not tests (I didn't look at the code yet). Thanks for clarifying.

unknwon commented:
Go's race detection output is pretty easy to reason:

Look for pairs of "Write at" and "Previous read at", for example, 
- https://buildkite.com/sourcegraph/sourcegraph/builds/46277#8d51fb5b-2238-42e6-b6af-343954de2deb/165-171
- https://buildkite.com/sourcegraph/sourcegraph/builds/46277#8d51fb5b-2238-42e6-b6af-343954de2deb/165-179

unknwon commented:
But I think there is one test failed: https://buildkite.com/sourcegraph/sourcegraph/builds/46277#8d51fb5b-2238-42e6-b6af-343954de2deb/247-248

rvantonder commented:
So this turned out not to be an obvious/easy fix. I tried creating local error declarations so that there isn't variable capture (this gets rid of the race indeed), but it also led to not assigning error values in the goroutine I pulled out. 

I learned the only way to guarantee assigning error values inside goroutines is with a channel. You could get away with not using a channel in certain conditions, but it doesn't work in the general case. Since we are interested in collecting errors across two+ concurrently executing routines here (indexed and unindexed), I now have this whole function only report errors via a dedicated error channel.

Right now the error check only checks the first error value read on the channel, though I think multiple errors can be reported with the new error channel changes. Will see if I can add a test that checks all reported errors (should multiple occur). But this is just a nice-to-have from my perspective, since in the previous logic the error result could have been non-deterministic anyway due capturing `error` over both go routines (I think).

rvantonder commented:
Aaannnnd looks like there's still a race. I'm not able to reliably reproduce locally.

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -555,95 +644,17 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
 	wg.Wait()
-	if err != nil {
+	close(errc)
+	if err := <-errc; err != nil {
```
Is it safe to assume we're going to have only one error at most?

```go
errc := make(chan error, 1)
```

I don't think it is safe because `searcherRepos` is a list, and each of the item in the list initiates a goroutine could send error to `errc`.

We should probably use [multierror](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+multierror&patternType=literal) in this case, then do a range over channel:

```go
for err := range <-errc {
	...
}
```

In fact, if there is a second goroutine tries to send error to `errc`, it blocks "forever", then our `wg.Wait()` blocks, until caller of this entire function gives up and potentially causes goroutine leak.

Disclaimer: I didn't run the code, it would be good to test this "more than one error returned" situation to verify my comments.

unknwon replied:
Demo to simulate the deadlock: https://play.golang.org/p/JEvl2cWNw7H

unknwon commented:
> Aaannnnd looks like there's still a race. I'm not able to reliably reproduce locally.

Hope you haven't closed your terminal... would you mind post the race output?

unknwon commented:
> But this is just a nice-to-have from my perspective, since in the previous logic the error result could have been non-deterministic anyway due capturing `error` over both go routines (I think).

I would say it is a bug... strictly speaking, we got lucky that we haven't encountered debugging pains/problems with losing errors in this code path. Maybe @slimsag have?

unknwon commented:
For this new race:

- Read at https://buildkite.com/sourcegraph/sourcegraph/builds/46334#87538c20-5cb1-4921-a637-34a6722a708c/88-626

https://github.com/sourcegraph/sourcegraph/blob/aeeba94afd54edfbe1ec564d0da706f3597b6cd7/cmd/frontend/graphqlbackend/textsearch.go#L630-L633

- Previous write at https://buildkite.com/sourcegraph/sourcegraph/builds/46334#87538c20-5cb1-4921-a637-34a6722a708c/88-630

https://github.com/sourcegraph/sourcegraph/blob/aeeba94afd54edfbe1ec564d0da706f3597b6cd7/cmd/frontend/graphqlbackend/textsearch.go#L654

My suggestion is adding a mutex around modifications to `common.` fields. But I want to know what @keegancsmith thinks.

rvantonder commented:
- [Buildkite race](https://buildkite.com/sourcegraph/sourcegraph/builds/46334#87538c20-5cb1-4921-a637-34a6722a708c/88-622)
- [Local race](https://gist.github.com/rvantonder/16d66fbaacc4ba3cdafab7682a30e6db)

rvantonder commented:
> I would say it is a bug... strictly speaking, we got lucky that we haven't encountered debugging pains/problems with losing errors in this code path.

Agree :) Would like to ship a fix with this.

keegancsmith commented:
Common is mutated, but needs to be done under a lock. I think rather try lift that function out totally, and make it behave more like zoektSearchHEAD

rvantonder commented:
Things are working now. PTAL. Here's what's changed:

> My suggestion is adding a mutex around modifications to common. (Joe)
> Common is mutated, but needs to be done under a lock. I think rather try lift that function out totally, and make it behave more like zoektSearchHEAD (Keegan)

I changed the function signature so it doesn't accept `common` and return it (this wasn't necessary). It now modifies `common` in the closure, as before, under a lock.

> We should probably use` multierror` in this case, then do a range over channel

Agree, and did this. I changed the test to use the error string value (not super enthused about this), but I don't see a way to test against the test type in the original test, since multierror wraps errors and doesn't expose the underlying list of errors.

rvantonder commented:
As for why I didn't lift out the function totally: there's a lot of state captured in the closure that would need more careful consideration than the current change. I'd rather not attempt it right now.

keegancsmith commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -555,94 +647,19 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
+	close(errc)
+	for e := range errc {
+		err = multierror.Append(err, e)
+	}
```
multierror will hide the cause of the failure from our error handling introspection. We shouldn't use it in this case.

keegancsmith commented on cmd/frontend/graphqlbackend/textsearch_test.go:
```diff
@@ -353,9 +353,14 @@ func TestSearchFilesInRepos(t *testing.T) {
 	_, _, err = searchFilesInRepos(context.Background(), args)
-	if !gitserver.IsRevisionNotFound(errors.Cause(err)) {
-		t.Fatalf("searching non-existent rev expected to fail with RevisionNotFoundError got: %v", err)
+	if err.Error() != want {
```
for example, this change in particular highlights why multi error is bad in our case. We want IsRevisionNotfound to return true.

rvantonder replied:
Yeah, as I said:

>  I changed the test to use the error string value (not super enthused about this) ...

I would've liked if a type 'like' multierr was possible without making `error` opaque, but it looks like this is not possible :)

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -553,98 +640,24 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
-		}(limitCtx, limitDone, repoRev)
+	if err := callSearcherOverRepos(searcherRepos); err != nil {
+		mu.Lock()
+		searchErr = err
```
Seems this is the only place used a mutex for `searchErr` (which as a local variable, is a noop?), we need lock/unlock for `searchErr` reads/writes inside goroutines, too.

keegancsmith replied:
all writes to searchErr have the mutex acquired. (there is more than one)

unknwon replied:
Ahhh... I'm blind!

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -502,6 +502,7 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
 		// TODO: convert wg to an errgroup
 		wg                sync.WaitGroup
 		mu                sync.Mutex
```
After @keegancsmith's commit, it now guards variable `searchErr`, what about renaming it to `searchErrMu`? Because it is not clear what things this mutex guards for by current name `mu`. WDYT?

keegancsmith replied:
it also guards common.

rvantonder commented:
![data-races](https://user-images.githubusercontent.com/888624/67603949-2bb02b00-f72f-11e9-9db2-3f96a38a3406.jpg)

Thanks @keegancsmith

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -553,98 +640,24 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
-		tr.LogFields(otlog.Object("searchErr", searchErr), otlog.Error(err), otlog.Bool("overLimitCanceled", overLimitCanceled))
-		if searchErr != nil && err == nil && !overLimitCanceled {
-			err = searchErr
+		tr.LogFields(otlog.Error(err), otlog.Error(err), otlog.Bool("overLimitCanceled", overLimitCanceled))
```
@keegancsmith It seems we have two `otlog.Error(err)` here, should we remove one of them?

keegancsmith replied:
good catch. It is benign but will remove one.

keegancsmith replied:
https://github.com/sourcegraph/sourcegraph/pull/6248
